### ISSUE TITLE
Moving `node_modules` (inside docker container) breaks analyzing other tools' dependencies

### ISSUE THREAD

**Issue by malice00** [oss-review-toolkit/ort#5174]
We have a react-native mobile project, which is initially setup using NPM. Then both Gradle (Android) and CocoaPods (iOS) reference several modules from the `node_modules` as their dependencies.
Now, because inside the docker container, the `node_modules` are move to another location, ORT fails to scan the dependencies for Gradle and CocoaPods.

android/build.gradle (snippet):
```
allprojects {
    repositories {
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            name "React Native @ Node-Modules"
            url "$rootDir/../node_modules/react-native/android"
        }
        maven {
            // Android JSC is installed from npm
            name "Android JSC @ Node-Modules"
            url "$rootDir/../node_modules/jsc-android/dist"
        }
    }
}
```

android/app/build.gradle (snippet):
```
apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
```

android/settings.gradle:
```
rootProject.name = 'MAAPP'
include ':react-native-config'
project(':react-native-config').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-config/android')
include ':react-native-default-preference'
project(':react-native-default-preference').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-default-preference/android')
include ':react-native-version-number'
project(':react-native-version-number').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-version-number/android')
include ':react-native-pdf'
project(':react-native-pdf').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-pdf/android')
include ':rn-fetch-blob'
project(':rn-fetch-blob').projectDir = new File(rootProject.projectDir, '../node_modules/rn-fetch-blob/android')
apply from: file("../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)
include ':app'
```

ios/Podfile (snippet):
```
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'
```

ios/Podfile.lock (snippet):
```
EXTERNAL SOURCES:
  DoubleConversion:
    :podspec: "../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec"
  FBLazyVector:
    :path: "../node_modules/react-native/Libraries/FBLazyVector"
  FBReactNativeSpec:
    :path: "../node_modules/react-native/React/FBReactNativeSpec"
  glog:
    :podspec: "../node_modules/react-native/third-party-podspecs/glog.podspec"
  Permission-Camera:
    :path: "../node_modules/react-native-permissions/ios/Camera.podspec"
```

Ort output:
```
17:45:26.337 [DefaultDispatcher-worker-4] INFO  org.ossreviewtoolkit.utils.common.DirectoryStash - Temporarily moving directory from '/project/node_modules' to '/project/stash16300321570829416566/node_modules'.
--- SNIP ---
17:45:30.813 [DefaultDispatcher-worker-5] INFO  org.ossreviewtoolkit.utils.common.ProcessCapture - Running 'pod spec which DoubleConversion --version=1.1.6 --allow-root --regex' in '/project/ios'...
17:45:33.434 [DefaultDispatcher-worker-5] ERROR org.ossreviewtoolkit.analyzer.managers.CocoaPods - Resolving CocoaPods dependencies for path 'ios/Podfile' failed with: IOException: Running 'pod spec which DoubleConversion --version=1.1.6 --allow-root --regex' in '/project/ios' failed with exit code 1:
[!] Can't find spec for DoubleConversion.
--- SNIP ---
17:45:48.874 [DefaultDispatcher-worker-6] ERROR org.ossreviewtoolkit.analyzer.managers.Gradle - Resolving Gradle dependencies for path 'android/build.gradle' failed with: BuildException: Could not fetch model of type 'DependencyTreeModel' using connection to Gradle distribution 'http://XXX/distributions/gradle-6.7-all.zip'.
Caused by: LocationAwareException: Settings file '/project/android/settings.gradle' line: 12
A problem occurred evaluating settings 'MAAPP'.
Caused by: GradleScriptException: A problem occurred evaluating settings 'MAAPP'.
Caused by: MissingResourceException: Could not read script '/project/node_modules/@react-native-community/cli-platform-android/native_modules.gradle' as it does not exist.
17:45:48.874 [DefaultDispatcher-worker-6] INFO  org.ossreviewtoolkit.analyzer.managers.Gradle - Resolving Gradle dependencies for path 'android/build.gradle' took 23.042516318s.
17:45:48.874 [DefaultDispatcher-worker-6] INFO  org.ossreviewtoolkit.analyzer.managers.Gradle - Resolving Gradle dependencies for path 'android/app/build.gradle'...
17:45:49.033 [DefaultDispatcher-worker-6] ERROR org.ossreviewtoolkit.analyzer.managers.Gradle - Resolving Gradle dependencies for path 'android/app/build.gradle' failed with: BuildException: Could not fetch model of type 'DependencyTreeModel' using connection to Gradle distribution 'http://XXX/distributions/gradle-6.7-all.zip'.
Caused by: LocationAwareException: Settings file '/project/android/settings.gradle' line: 12
A problem occurred evaluating settings 'MAAPP'.
Caused by: GradleScriptException: A problem occurred evaluating settings 'MAAPP'.
Caused by: MissingResourceException: Could not read script '/project/node_modules/@react-native-community/cli-platform-android/native_modules.gradle' as it does not exist.
17:45:49.033 [DefaultDispatcher-worker-6] INFO  org.ossreviewtoolkit.analyzer.managers.Gradle - Resolving Gradle dependencies for path 'android/app/build.gradle' took 158.727381ms.
```

Now I know the error from CocoaPods is a little misleading, as it would also come because CP itself doesn't correctly resolve a file included from the file-system, however the problem would still be that the referenced file doesn't exist (anymore / yet).

I guess the question is if it is really necessary to move the `node_modules` directory, or if it is maybe possible to 'postpone' analyzing Gradle & CocoaPods while NPM is still running.

**Comment by sschuberth** [comment]
I believe there are at least two issues that come into play here:

1. Definition files for different package managers are actually [analyzed in parallel](https://github.com/oss-review-toolkit/ort/blob/58a5104ee9e9b9a35334f85c8cfcdf76112356e8/analyzer/src/main/kotlin/Analyzer.kt#L127-L168).
2. In any case, the order in which package managers are requested to analyze their definition files is more or less undefined.

As a simple way to address this, I believe that we cloud change the analyzer's `-m` / `--package-managers` option to that, if specified, the order of the given package managers is maintained for running them, plus we could introduce a new option to disable parallel analysis.

Both things together should solve your issue as you could then say `ort analyze --no-parallelism --package-managers npm,gradle,cocoapods`. What do you think?

**Comment by malice00** [comment]
That sounds like a good solution!
Does turning off parallelism have much effect on eg speed? In this case, I don't mind parallelism for gradle and cocoapods, the only important thing is that NPM is finished... Then again, that's probably a lot harder to implement, especially if someone comes along with groups of managers that are dependent. I guess all or nothing is the way to go here.

**Comment by sschuberth** [comment]
Sorry for the late reply.

> Does turning off parallelism have much effect on eg speed?

We assume yes, but it of course depends a bit on whether I/O is the bottleneck on the machine you're running ORT on or not.

The new idea @mnonnenmacher and I came up with is to not disable parallelism completely, but to have a mechanism telling ORT's analyzer which package manager depends on what other package manager (similar to [Gradle's `mustRunAfter` for tasks](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:ordering_tasks)). But as the CLI syntax would get a bit complex for that, such an option should probably go into `ort.conf` / `.ort.yml`.

**Comment by mnonnenmacher** [comment]
@malice00 I have added a configuration option in #5465 that makes it possible to configure that one package manager must run after another. Could you give this a try and report back if it fixes your issue if you configure Gradle and CocoaPods to run after NPM?

**Comment by malice00** [comment]
@mnonnenmacher Works perfectly!

**Comment by mnonnenmacher** [comment]
@malice00 Thanks for the feedback!
